beater/pub.go

```diff
@@ -102,6 +115,34 @@ func (p *publisher) Send(req pendingReq) error {
 
 func (p *publisher) run() {
 	for req := range p.pendingRequests {
-		p.client.PublishAll(req.payload.Transform(req.config))
+		p.processPendingReq(req)
+	}
+}
+
+func (p *publisher) processPendingReq(req pendingReq) {
+	var tx *elasticapm.Transaction
+	maybeStartSpan := func(name, spanType string) *elasticapm.Span {
+		if tx != nil {
```

Comment:
what do you think about adding nil checks in the methods themselves in the agent so that users don't have to worry whether there is a transaction going on or not?

Reply:
The problem is that part of the API involves fields on the transaction, which means doing something like

```go
tx.Result = "foo"
```

will panic if tx is nil. I do agree it's not great as it is, though. The alternative is that we unexpose all fields and replace them with methods. That might be better. On the other hand, forcing the nil check ensures that the instrumented application avoids overhead as much as possible.

Reply:
It might be reasonable to do a limited amount of nil checking (maybe on Transaction.StartSpan), but I think it would be more useful to force callers to check the result by returning a second bool result (the "comma-ok" pattern), like:

```go
// tracer.StartTransaction will return (nil, false) if the tracer is inactive or the transaction's name is being filtered
tx, ok := tracer.StartTransaction("name", "type")
if ok {
    defer tx.Done(-1)
    tx.Context.Set...
}

// tx.StartSpan will return (nil, false) if tx is nil, if the transaction is not  being sampled, or if the transaction's max spans limit has been reached.
span, ok := tx.StartSpan("name", "type", parentSpan)
if ok {
    defer span.Done(-1)
    span.Context.Set...
}
```

We can also make the Done methods no-ops for nil Transactions and Spans, but I'm a bit wary about that. A developer seeing that might think it's also reasonable to use the object's fields without checking the value for nil or the bool result.

Alternatively, I'm considering adding a new helper `func DoSpan(ctx context.Context, name, spanType string, f func(ctx context.Context))`. This would take care of checking if the span is nil, and calling its Done method, so you could do something like this:

```go
ctx := context.Background()
if req.trace {
    tx, ok := tracer.StartTransaction("name", "type")
    if ok {
        defer tx.Done(-1)
        ctx = elasticapm.ContextWithTransaction(ctx, tx)
    }
}

var events []beat.Event
elasticapm.DoSpan(ctx, "Transform", "Publisher", func() {
    events = req.payload.Transform(req.config)
})
```

DoSpan will be defined something like:

```go
func DoSpan(ctx context.Context, name, spanType string, f func(context.Context)) {
    span, ctx := elasticapm.StartSpan(name, spanType)
    if span != nil {
        defer span.Done(-1)
    }
    f(ctx)
}
```

Reply:
I've just created https://github.com/elastic/apm-agent-go/pull/92, which is the other other alternative: never create nil spans. StartTransaction already never returns nil transactions. This simplifies the code here to:

```go
var tx *elasticapm.Transaction
if req.trace {
    tx = tracer.StartTransaction("name", "type")
    defer tx.Done(-1)
}

span := tx.StartSpan("Transform", "Publisher", nil)
events := req.payload.Transform(req.config)
span.Done(-1)
```